Firestore/core/src/core/sync_engine.h

```diff
@@ -220,6 +219,53 @@ class SyncEngine : public remote::RemoteStoreCallback, public QueryEventSource {
     bool document_received = false;
   };
 
+  class LimboResolutionQueue {
+   public:
+    LimboResolutionQueue() = default;
+    LimboResolutionQueue(const LimboResolutionQueue&) = delete;
+
+    void push_back(const model::DocumentKey& key);
+    model::DocumentKey front() const;
+    void pop_front();
+    void remove(const model::DocumentKey& key);
+
+    bool empty() const {
+      return queue_.empty();
+    }
+
+    bool contains(const model::DocumentKey& key) const;
+    std::vector<model::DocumentKey> keys() const;
+
+   private:
+    class QueueEntry {
```

Comment:
Since this wrapper only seems to exist to support cancellation, I was wondering if you are up for a challenge. You could advance to the stage of a "true hacker" (tm) if you use `unordered_map` to indicate whether an entry is still active. You can remove inactive entries from the unordered_map and clean up the vector by treating the map as a source of truth.